src/model_repository_manager.cc

```diff
@@ -58,6 +58,27 @@ AddToSet(const std::set<T>& src, std::set<T>* dest)
   }
 }
 
+template <class Mutex>
+class RevLockGuard {
+ public:
+  RevLockGuard(Mutex& mu, bool op = true) : mu_(mu), op_(op)
+  {
+    if (op_)
+      mu_.unlock();
+  }
+  ~RevLockGuard()
+  {
+    if (op_)
+      mu_.lock();
+  }
+  RevLockGuard(const RevLockGuard&) = delete;
+  RevLockGuard& operator=(const RevLockGuard&) = delete;
+
+ private:
+  Mutex& mu_;
+  bool op_;
+};
```

Comment:
It is because the lock scope span across different nested function calls. It needs to be held at the outer function and then release and re-held at an inner function.

Reply:
See another comment below, I think this nested locking can be avoided if you clean up the logic. And you should consider declare additional mutexes for finer control, the use of `poll_mtx` is everywhere and the chance of over-use is high.

Reply:
Yes.
Update commit: [Refactor parallel load logic](https://github.com/triton-inference-server/core/pull/165/commits/0fa4e296b613806d0bd7de77775f09e51d866d21)](https://github.com/triton-inference-server/core/pull/165/commits/0fa4e296b613806d0bd7de77775f09e51d866d21)